airbyte-integrations/connectors/source-vertica/source_vertica/source.py

```diff
@@ -0,0 +1,173 @@
+            A stream is an AirbyteStream object that includes:
+            - its stream name (or table name in the case of Postgres)
+            - json_schema providing the specifications of expected schema for this stream (a list of columns described
+            by their names and types)
+        """
+        streams = []
+
+        connection = self.create_connection(config)
+
+        tables = self.get_tables(logger, config, connection)
+
+        for table in tables:
+            properties = {}
+            columns = self.get_columns(logger, config, connection, table)
+
+            for column in columns:
+                properties[column[0]] = {
+                    "type": column[1]
+                }
+
+            streams.append(
+                AirbyteStream(
+                    name=table,
+                    json_schema={
+                        "$schema": "http://json-schema.org/draft-07/schema#",
+                        "type": "object",
+                        "properties": properties,
+                    }
+                )
+            )
+
+        return AirbyteCatalog(streams=streams)
+
+    def read(
+        self, logger: AirbyteLogger, config: json, catalog: ConfiguredAirbyteCatalog, state: Dict[str, any]
+    ) -> Generator[AirbyteMessage, None, None]:
+        """
+        Returns a generator of the AirbyteMessages generated by reading the source with the given configuration,
+        catalog, and state.
+
+        :param logger: Logging object to display debug/info/error to the logs
+            (logs will not be accessible via airbyte UI if they are not passed to this logger)
+        :param config: Json object containing the configuration of this source, content of this json is as specified in
+            the properties of the spec.yaml file
+        :param catalog: The input catalog is a ConfiguredAirbyteCatalog which is almost the same as AirbyteCatalog
+            returned by discover(), but
+        in addition, it's been configured in the UI! For each particular stream and field, there may have been provided
+        with extra modifications such as: filtering streams and/or columns out, renaming some entities, etc
+        :param state: When a Airbyte reads data from a source, it might need to keep a checkpoint cursor to resume
+            replication in the future from that saved checkpoint.
+            This is the object that is provided with state from previous runs and avoid replicating the entire set of
+            data everytime.
+
+        :return: A generator that produces a stream of AirbyteRecordMessage contained in AirbyteMessage object.
+        """
+        connection = self.create_connection(config)
+
+        for stream in catalog.streams:
+            stream_name = stream.stream.name
+
+            query = f"SELECT * FROM {stream_name}"
+            results, columns = self.query(logger, config, connection, query)
+
+            for row in results:
```

Comment:
Are you sure that you can directly map the db result set to a valid JSON file conforming to the above defined schema on discover?

Reply:
I didn't quite understand this. Is this the same concern you had under the third point of your second comment?

Reply:
that is correct. the mapping here needs to conform to the json schema.

Reply:
I have created a mapper function.

Could you please tell me if this output (discover) looks correct,
`{"type": "CATALOG", "catalog": {"streams": [{"name": "customer_dimension", "json_schema": {"$schema": "http://json-schema.org/draft-07/schema#", "type": "object", "properties": {"customer_key": {"type": "integer"}, "customer_type": {"type": "string"}, "customer_name": {"type": "string"}, "customer_gender": {"type": "string"}, "title": {"type": "string"}, "household_id": {"type": "integer"}, "customer_address": {"type": "string"}, "customer_city": {"type": "string"}, "customer_state": {"type": "string"}, "customer_region": {"type": "string"}, "marital_status": {"type": "string"}, "customer_age": {"type": "integer"}, "number_of_children": {"type": "integer"}, "annual_income": {"type": "integer"}, "occupation": {"type": "string"}, "largest_bill_amount": {"type": "integer"}, "store_membership_card": {"type": "integer"}, "customer_since": {"type": "string", "format": "date"}, "deal_stage": {"type": "string"}, "deal_size": {"type": "integer"}, "last_deal_update": {"type": "string", "format": "date"}}}}, {"name": "date_dimension", "json_schema": {"$schema": "http://json-schema.org/draft-07/schema#", "type": "object", "properties": {"date_key": {"type": "integer"}, "date": {"type": "string", "format": "date"}, "full_date_description": {"type": "string"}, "day_of_week": {"type": "string"}, "day_number_in_calendar_month": {"type": "integer"}, "day_number_in_calendar_year": {"type": "integer"}, "day_number_in_fiscal_month": {"type": "integer"}, "day_number_in_fiscal_year": {"type": "integer"}, "last_day_in_week_indicator": {"type": "integer"}, "last_day_in_month_indicator": {"type": "integer"}, "calendar_week_number_in_year": {"type": "integer"}, "calendar_month_name": {"type": "string"}, "calendar_month_number_in_year": {"type": "integer"}, "calendar_year_month": {"type": "string"}, "calendar_quarter": {"type": "integer"}, "calendar_year_quarter": {"type": "string"}, "calendar_half_year": {"type": "integer"}, "calendar_year": {"type": "integer"}, "holiday_indicator": {"type": "string"}, "weekday_indicator": {"type": "string"}, "selling_season": {"type": "string"}}}}, {"name": "employee_dimension", "json_schema": {"$schema": "http://json-schema.org/draft-07/schema#", "type": "object", "properties": {"employee_key": {"type": "integer"}, "employee_gender": {"type": "string"}, "courtesy_title": {"type": "string"}, "employee_first_name": {"type": "string"}, "employee_middle_initial": {"type": "string"}, "employee_last_name": {"type": "string"}, "employee_age": {"type": "integer"}, "hire_date": {"type": "string", "format": "date"}, "employee_street_address": {"type": "string"}, "employee_city": {"type": "string"}, "employee_state": {"type": "string"}, "employee_region": {"type": "string"}, "job_title": {"type": "string"}, "reports_to": {"type": "integer"}, "salaried_flag": {"type": "integer"}, "annual_salary": {"type": "integer"}, "hourly_rate": {"type": "number"}, "vacation_days": {"type": "integer"}}}}, {"name": "inventory_fact", "json_schema": {"$schema": "http://json-schema.org/draft-07/schema#", "type": "object", "properties": {"date_key": {"type": "integer"}, "product_key": {"type": "integer"}, "product_version": {"type": "integer"}, "warehouse_key": {"type": "integer"}, "qty_in_stock": {"type": "integer"}, "inventory_date": {"type": "string", "format": "date"}}}}, {"name": "product_dimension", "json_schema": {"$schema": "http://json-schema.org/draft-07/schema#", "type": "object", "properties": {"product_key": {"type": "integer"}, "product_version": {"type": "integer"}, "product_description": {"type": "string"}, "sku_number": {"type": "string"}, "category_description": {"type": "string"}, "department_description": {"type": "string"}, "package_type_description": {"type": "string"}, "package_size": {"type": "string"}, "fat_content": {"type": "integer"}, "diet_type": {"type": "string"}, "weight": {"type": "integer"}, "weight_units_of_measure": {"type": "string"}, "shelf_width": {"type": "integer"}, "shelf_height": {"type": "integer"}, "shelf_depth": {"type": "integer"}, "product_price": {"type": "integer"}, "product_cost": {"type": "integer"}, "lowest_competitor_price": {"type": "integer"}, "highest_competitor_price": {"type": "integer"}, "average_competitor_price": {"type": "integer"}, "discontinued_flag": {"type": "integer"}}}}, {"name": "promotion_dimension", "json_schema": {"$schema": "http://json-schema.org/draft-07/schema#", "type": "object", "properties": {"promotion_key": {"type": "integer"}, "promotion_name": {"type": "string"}, "price_reduction_type": {"type": "string"}, "promotion_media_type": {"type": "string"}, "ad_type": {"type": "string"}, "display_type": {"type": "string"}, "coupon_type": {"type": "string"}, "ad_media_name": {"type": "string"}, "display_provider": {"type": "string"}, "promotion_cost": {"type": "integer"}, "promotion_begin_date": {"type": "string", "format": "date"}, "promotion_end_date": {"type": "string", "format": "date"}}}}, {"name": "shipping_dimension", "json_schema": {"$schema": "http://json-schema.org/draft-07/schema#", "type": "object", "properties": {"shipping_key": {"type": "integer"}, "ship_type": {"type": "string"}, "ship_mode": {"type": "string"}, "ship_carrier": {"type": "string"}}}}, {"name": "vendor_dimension", "json_schema": {"$schema": "http://json-schema.org/draft-07/schema#", "type": "object", "properties": {"vendor_key": {"type": "integer"}, "vendor_name": {"type": "string"}, "vendor_address": {"type": "string"}, "vendor_city": {"type": "string"}, "vendor_state": {"type": "string"}, "vendor_region": {"type": "string"}, "deal_size": {"type": "integer"}, "last_deal_update": {"type": "string", "format": "date"}}}}, {"name": "vmart_load_success", "json_schema": {"$schema": "http://json-schema.org/draft-07/schema#", "type": "object", "properties": {"a": {"type": "integer"}}}}, {"name": "warehouse_dimension", "json_schema": {"$schema": "http://json-schema.org/draft-07/schema#", "type": "object", "properties": {"warehouse_key": {"type": "integer"}, "warehouse_name": {"type": "string"}, "warehouse_address": {"type": "string"}, "warehouse_city": {"type": "string"}, "warehouse_state": {"type": "string"}, "warehouse_region": {"type": "string"}}}}]}}`